### ISSUE TITLE
Experiment __init__() + setup() + configure() is a confusing mess

### ISSUE THREAD

**Issue by jessesnyder** [Dallinger/Dallinger#3125]
## Summary
The interactions between this cluster of API methods is confusing, and demos don't provide clarifying examples. The ideas here are from a brainstorming session between @alecpm and @jessesnyder.

## Goals

1. Custom experiments should not need to override `__init__()`, except under extraordinary circumstances
2. Custom experiments should not need to call `super()` in standard cases, since this is always confusing and it can be hard to tell when it matters
3. Base class should handle the database transaction in standard cases

## Tentative Proposal

1. Experiment base class's `__init__()` method calls a private `_setup()` (name TBD) and finishes with a database commit
2. `_setup()` handles idempotence by checking `if self.networks() is None` and creates basic network as the current base class's `setup()` method does
3. `_setup()` calls `self.setup()` at the end
4. `setup()` has an empty implementation in the base class, and is where the subclass adds its own special nodes to the now-initialized networks
5. statements like `self.initial_recruitment_size = 2` belong in `configure()`, not `__init__()` or `setup()`

**Comment by jessesnyder** [comment]
tag @jacobyn @pmcharrison as I'm curious to hear your thoughts.

**Comment by pmcharrison** [comment]
Thanks @jessesnyder - yes, we have struggled with this too. Your proposal sounds sensible. In practice, we do end up wanting to insert code both at the one-time setup phase (e.g. network creation) and at the object instantiation phase (which happens multiple times). It is not particularly elegant having `self.networks() is not None` as the definition of whether the setup has happened already (what if you don't want to create any networks straight away?) but I think if you wrap that in a sensible `property` then the code should be perfectly readable.

**Comment by alecpm** [comment]
@pmcharrison So something like an `is_setup_complete` boolean property on the experiment whose default implementation checks if networks have been created? That sounds reasonable to me, though I'm getting hesitant to add even more APIs to Dallinger if we can avoid them. Is the case where someone needs an experiment that doesn't setup networks initially is a common one? If not, perhaps that's a situation where the experiment creator would need a custom `__init__`?

**Comment by pmcharrison** [comment]
Hi @alecpm, don't worry about it if you don't like it, it's really a small thing!

**Comment by pmcharrison** [comment]
Maybe related to this is a confusion that I have while working on this part of the code. The `Experiment` class can be initialised with `session=None` OR with `session=dallinger.db.session`. It's not very clear to me in what cases the former should happen, and when the latter should happen. It might be good to clarify the behaviour here.

**Comment by alecpm** [comment]
@pmcharrison Generally the experiment class is only instantiated when programmatically launching experiment runs using APIs like `Experiment.run` and `Experiment.collect`, and in tests. By default when an experiment is instantiated without a session, the `setup` and `configure` methods aren't run. The Experiment should always be instantiated with a DB session when used by the experiment server.

**Comment by pmcharrison** [comment]
@alecpm Makes sense. Maybe there is a way as part of this breaking change to make this dual functionality more explicit in the Experiment class.
